src/net.cpp

```diff
@@ -2331,6 +2332,11 @@ bool CAddrDB::Read(CAddrMan& addr)
 
     CDataStream ssPeers(vchData, SER_DISK, CLIENT_VERSION);
 
+    return Read(addr, ssPeers, hashIn);
+}
+
+bool CAddrDB::Read(CAddrMan& addr, CDataStream& ssPeers, uint256& hashIn)
+{
     // verify stored checksum matches input data
     uint256 hashTmp = Hash(ssPeers.begin(), ssPeers.end());
```

Comment:
My thinking with passing hashIn as a parameter was that it would allow me to write a test on the "inner read" to verify an error is thrown if hashIn is not the same as the hash of ssPeers. Something like:

```c++ 
BOOST_AUTO_TEST_CASE(caddrdb_hash)
{
    CAddrManUncorrupted addrmanUncorrupted;
    CService addr1 = CService("250.7.1.1", 8333);
    addrmanUncorrupted.Add(CAddress(addr1), CService("252.5.1.1", 8333));

    // Test that the de-serialization does not throw an exception.
    CDataStream ssPeers1 = AddrmanToStream(addrmanUncorrupted);

    addrmanUncorrupted.Add(CAddress(addr3), CService("253.3.3.33", 8333));
    CDataStream ssPeersDiff = AddrmanToStream(addrmanUncorrupted);
    uint256 badHashIn = Hash(ssPeersDiff.begin(), ssPeersDiff.end());

    CAddrDB adb;
    CAddrMan addrman;
    adb.Read(addrman, ssPeers2, badHashIn); // Assert this throws an exception since the hashes don't match
}
```
but I decided to wait until I had more time to work on it.

How interested are you in a system for mocking out filesystem objects rather than having to do everything at the stream level?